core/src/tx.rs

```diff
@@ -195,46 +194,66 @@ impl AcceptedTransaction {
             return Err(reason);
         }
 
-        for instruction in &self.payload.instructions {
-            instruction
-                .clone()
-                .execute(account_id.clone(), &wsv_temp)
-                .map_err(|reason| InstructionExecutionFail {
-                    instruction: instruction.clone(),
-                    reason: reason.to_string(),
-                })
-                .map_err(TransactionRejectionReason::InstructionExecution)?;
-
-            // Permission validation is skipped for genesis.
-            if !is_genesis {
-                #[cfg(feature = "roles")]
-                {
-                    let instructions = permissions::unpack_if_role_grant(instruction.clone(), wsv)
+        // WSV is cloned here so that instructions don't get applied to the blockchain
+        // Therefore, this instruction execution validates before actually executing
+        let wsv_temp = wsv.clone();
+        match &self.payload.instructions {
+            Executable::Instructions(instructions) => {
```

Comment:
What do you think about moving the branch for instructions and the branch for wasm into separate functions?

It will probably make code more readable here.

Reply:
I've restructured this method a bit to make it more readable